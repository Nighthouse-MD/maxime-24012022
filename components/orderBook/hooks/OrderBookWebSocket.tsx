import WebSocketOutMessage from "./WebSocketOutMessage";
import Constants from "../../../Constants";

export default class OrderBookWebSocket {
    private url: string;
    private productId: string;
    private ws: WebSocket;
    private onMessageHandler;
    private onError;

    constructor(url: string, productId: string, onMessageHandler, onError) {
        this.url = url;
        this.productId = productId;
        this.onMessageHandler = onMessageHandler;
        this.onError = onError;

        this.init();
    }

    public init = () => {
        if (!this.ws || (this.ws.readyState !== WebSocket.OPEN && this.ws.readyState !== WebSocket.CONNECTING)) {
            this.ws = new WebSocket(this.url);
            this.ws.onopen = () => {
                const subEvent = new WebSocketOutMessage("subscribe", this.productId, Constants.WEBSOCKET_FEED);
                this.ws.send(subEvent.toString());
                console.log('WS connection opened')
            };

            this.ws.onmessage = this.onMessageHandler;

            this.ws.onerror = (event) => {
                this.onError();
                console.log(`There was an error with the websocket: ${event}`);
            };
        }
    }

    public close = () => {
        if (this.ws && this.ws.readyState !== WebSocket.CLOSED && this.ws.readyState !== WebSocket.CLOSING) {
            if (this.ws.readyState !== WebSocket.CONNECTING) {
                const unsubEvent = new WebSocketOutMessage("unsubscribe", this.productId, Constants.WEBSOCKET_FEED);
                this.ws.send(unsubEvent.toString())
            }
            this.ws.close();
            console.log('WS connection closed')
        }
    }
}

