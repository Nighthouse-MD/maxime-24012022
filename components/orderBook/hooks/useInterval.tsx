import { useEffect, useLayoutEffect, useRef } from 'react'
export const isBrowser = typeof window !== "undefined";

function useInterval(callback: () => void, delay: number | null) {
    const savedCallback = useRef(callback)

    // if (isBrowser)
    // Remember the latest callback if it changes.
    useLayoutEffect(() => {
        savedCallback.current = callback
    }, [callback])

    // Set up the interval.
    useEffect(() => {
        // Don't schedule if no delay is specified.
        // Note: 0 is a valid value for delay.
        if (!delay && delay !== 0) {
            return
        }

        const id = setInterval(() => savedCallback.current(), delay)

        return () => clearInterval(id)
    }, [delay])
}

export default useInterval
