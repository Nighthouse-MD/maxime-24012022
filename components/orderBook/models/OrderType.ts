enum OrderType {
    ask = "sell",
    bid = "buy"
}

export default OrderType;