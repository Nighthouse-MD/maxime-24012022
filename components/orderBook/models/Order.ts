import OrderType from "./OrderType";

class Order {
    type: OrderType;
    price: number;
    quantity: number;
    cumulativeVolume: number;
    volumePercentage: number;

    constructor(type: OrderType, price: number, quantity: number) {
        this.type = type;
        this.price = price;
        this.quantity = quantity;
        this.cumulativeVolume = 0;
        this.volumePercentage = 0;
    }
}

export default Order;
