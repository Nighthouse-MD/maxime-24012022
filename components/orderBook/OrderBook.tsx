import { useState, useMemo, useCallback } from 'react';
import { Row, Col, Container } from 'react-bootstrap';
import styles from './OrderBook.module.scss'

import OrderRow from './subcomponents/OrderRow';
import OrderRowNumbers from './subcomponents/OrderRowNumbers';
import OrderBookSettings from './subcomponents/OrderBookSettings';
import OrderBookHeader from './subcomponents/OrderBookHeader';
import FeedToggle from './subcomponents/FeedToggle';

import Order from './models/Order';

import { orderByPropertyNameDesc } from '../../tools/orderBy';
import { addCumulativeVolumes, addVolumePercentage, calculateSpreadInfo } from './OrderBookHelpers';

import Hooks from './hooks';
import CommonHooks from '../common/hooks';
import OrderType from './models/OrderType';
import ReconnectModal from './subcomponents/ReconnectModal';

const OrderBook = ({ defaultDepth, throttle }: Props) => {
  const [depth, setDepth] = useState<number>(defaultDepth);
  const [renderThrottleInterval, setRenderThrottleInterval] = useState<number>(throttle);
  const throttleTick = CommonHooks.useThrottleTick(renderThrottleInterval);
  const { asks, bids, pair, wasDisconnected, handleReconnect, handleFeedToggle } = Hooks.useOrderBookFeed();

  const throttledOrderBookRows = useMemo(
    () => {
      if (!asks.length && !bids.length)
        return <Row>
          <Col className={styles.noConnection}>No websocket connection</Col>
        </Row>;

      const composeOrderRows = () => {

        //order and take depth
        const asksToShow = orderByPropertyNameDesc(asks, 'price').slice(Math.max(asks.length - depth, 0));
        const bidsToShow = orderByPropertyNameDesc(bids, 'price').slice(0, depth);

        //calculate spread
        const spread = calculateSpreadInfo(asksToShow, bidsToShow);

        //add cumulative volumes
        const asksWithVolumes = addCumulativeVolumes(asksToShow, false, depth);
        const bidsWithVolumes = addCumulativeVolumes(bidsToShow, true, depth);

        //get the biggest cumulative volume to define 100 %
        const biggestVolume = asksWithVolumes.length && bidsWithVolumes.length ? (asksWithVolumes[0].cumulativeVolume > bidsWithVolumes[bidsWithVolumes.length - 1].cumulativeVolume ?
          asksWithVolumes[0].cumulativeVolume :
          bidsWithVolumes[bidsWithVolumes.length - 1].cumulativeVolume) : 0;

        //add volume percentage for graph rendering
        const asksWithVolumePercentage = addVolumePercentage(asksWithVolumes, biggestVolume);
        const bidsWithVolumePercantage = addVolumePercentage(bidsWithVolumes, biggestVolume);

        //create OrderRow components
        const mapToOrderRow = (o: Order) =>
          <OrderRow
            key={o.price}
            type={o.type}
            volumePercentage={o.volumePercentage}>
            <OrderRowNumbers
              type={o.type}
              price={o.price.toFixed(2).toLocaleString()}
              quantity={o.quantity.toLocaleString()}
              cumulativeVolume={o.cumulativeVolume.toLocaleString()} />
          </OrderRow>;

        const askComponents = asksWithVolumePercentage.map((o: Order) => mapToOrderRow(o));
        const bidComponents = bidsWithVolumePercantage.map((o: Order) => mapToOrderRow(o));

        return { askComponents, bidComponents, spread };
      }

      const { askComponents, bidComponents, spread } = composeOrderRows();

      return <Row>
        <Col
          xs={{ span: 12, order: 1 }}
          md={{ span: 6, order: 3 }}>
          <OrderBookHeader type={OrderType.ask} />
          <div className={styles.responsiveAskRows}>
            {askComponents}
          </div>
        </Col>
        <Col
          xs={{ span: 12, order: 2 }}
          md={{ span: 12, order: 1 }}
          className={`${styles.spread} text-center`}>
          Spread: {spread.amount.toFixed(1)} ({spread.percent}%)
        </Col>
        <Col
          xs={{ span: 12, order: 3 }}
          md={{ span: 6, order: 2 }}>
          <OrderBookHeader type={OrderType.bid} />
          <div className="d-flex flex-column">
            {bidComponents}
          </div>
        </Col>
      </Row>;
    },
    [throttleTick]
  );

  return (
    <Container className={styles.orderBookContainer}>
      <ReconnectModal show={wasDisconnected} handleReconnect={handleReconnect} />
      <Row>
        <Col>
          <h2>{pair.substring(3)}</h2>
        </Col>
      </Row>
      <Row>
        <Col>
          <OrderBookSettings
            depth={depth}
            handleDepthChange={useCallback((e) => setDepth(parseInt(e.target.value)), [])}
            throttleInterval={renderThrottleInterval}
            handleThrottleIntervalChange={useCallback((e) => setRenderThrottleInterval(parseInt(e.target.value)), [])}
          />
        </Col>
      </Row>
      <hr />
      <br />
      <Row>
        <Col className={styles.orderBookTitle}>
          Order book
        </Col>
      </Row>
      <Row>
        <Col >
          {throttledOrderBookRows}
        </Col>
      </Row>
      <br />
      <Row>
        <Col>
          <FeedToggle handleFeedToggle={handleFeedToggle} />
        </Col>
      </Row>
    </Container>
  );
}

interface Props {
  defaultDepth: number;
  throttle: number;
}

export default OrderBook;