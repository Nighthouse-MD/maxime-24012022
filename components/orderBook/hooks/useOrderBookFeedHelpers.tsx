import Order from "../models/Order";

const addOrdersToList = (existingOrders: Order[], newOrders: Order[], resetList: boolean) => {
    let updatedOrders = resetList ? [] : [...existingOrders];
    newOrders.forEach(order => {
        updatedOrders = [...updatedOrders.filter(x => x.price !== order.price)];
        if (order.quantity > 0) {
            updatedOrders.push(order);
        }
    });
    return updatedOrders;
}

export { addOrdersToList };