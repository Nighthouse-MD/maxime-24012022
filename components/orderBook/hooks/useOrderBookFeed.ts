import { useState, useEffect, useCallback } from 'react';
import Order from '../models/Order';
import OrderType from '../models/OrderType';
import CONSTANTS from '../../../Constants';
import OrderBookWebSocket from './OrderBookWebSocket';
import { addOrdersToList } from './useOrderBookFeedHelpers';

const useOrderBookFeed = () => {
    const [asks, setAsks] = useState<Order[]>([]);
    const [bids, setBids] = useState<Order[]>([]);
    const [pair, setPair] = useState<string>("PI_XBTUSD");
    const [wasDisconnected, setWasDisconnected] = useState<boolean>(false);
    const [handleReconnect, setHandleReconnect] = useState<() => void>();

    const handleFeedToggle = useCallback((e) => {
        e.target.blur();
        setAsks([]);
        setBids([]);
        setPair(pair => pair === "PI_XBTUSD" ? "PI_ETHUSD" : "PI_XBTUSD")
    }, []);

    useEffect(() => {
        const onMessageHandler = (event) => {
            try {
                const eventData = JSON.parse(event.data);

                if (!!eventData.asks && !!eventData.bids) {
                    const isSnapshotData = eventData.feed === CONSTANTS.WEBSOCKET_SNAPSHOT_FEED;
                    const newAsks = eventData.asks.map((ask: number[]) => new Order(OrderType.ask, ask[0], ask[1]));
                    const newBids = eventData.bids.map((bid: number[]) => new Order(OrderType.bid, bid[0], bid[1]));

                    setAsks(existingAsks => addOrdersToList(existingAsks, newAsks, isSnapshotData));
                    setBids(existingBids => addOrdersToList(existingBids, newBids, isSnapshotData));
                }
            }
            catch (event) {
                console.log("There was an error with the websocket data parsing.");
                setWasDisconnected(true);
            }
        };

        const obws = new OrderBookWebSocket(CONSTANTS.WEBSOCKET_URL, pair, onMessageHandler, () => {
            setWasDisconnected(true);
        });

        window.onblur = () => {
            obws.close();
            setWasDisconnected(true);
        }

        window.onbeforeunload = () => {
            obws.close();
            setWasDisconnected(true);
        }

        const reconnect = () => {
            obws.init();
            setWasDisconnected(false);
        }

        setHandleReconnect(() => { return reconnect; });

        return () => obws.close();
    }, [pair]);

    return { asks, bids, pair, wasDisconnected, handleReconnect, handleFeedToggle };
}

export default useOrderBookFeed;