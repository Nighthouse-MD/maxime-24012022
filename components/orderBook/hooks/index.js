import useOrderBookFeed from "./useOrderBookFeed";

export default {
    useOrderBookFeed
}