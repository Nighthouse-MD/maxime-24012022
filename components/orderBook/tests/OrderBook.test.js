import React from 'react';
import { render, screen } from '@testing-library/react';
import { } from '@testing-library/react-hooks'
import { jest } from '@jest/globals'

import OrderBook from '../OrderBook';
import Order from '../models/Order';
import OrderType from '../models/OrderType';
import hooks from '../hooks';

const asks = [
    new Order(OrderType.ask, 30, 3),
    new Order(OrderType.ask, 29, 3),
    new Order(OrderType.ask, 28, 3),
    new Order(OrderType.ask, 27, 3),
    new Order(OrderType.ask, 26, 3),
    new Order(OrderType.ask, 25, 3),
    new Order(OrderType.ask, 24, 3),
    new Order(OrderType.ask, 23, 3),
    new Order(OrderType.ask, 22, 3),
    new Order(OrderType.ask, 21, 3),
    new Order(OrderType.ask, 20, 3),
    new Order(OrderType.ask, 19, 3),
    new Order(OrderType.ask, 18, 3),
    new Order(OrderType.ask, 17, 3),
    new Order(OrderType.ask, 16, 3)
];

const bids = [
    new Order(OrderType.bid, 15, 3),
    new Order(OrderType.bid, 14, 3),
    new Order(OrderType.bid, 13, 3),
    new Order(OrderType.bid, 12, 3),
    new Order(OrderType.bid, 11, 3),
    new Order(OrderType.bid, 10, 3),
    new Order(OrderType.bid, 9, 3),
    new Order(OrderType.bid, 8, 3),
    new Order(OrderType.bid, 7, 3),
    new Order(OrderType.bid, 6, 6),
    new Order(OrderType.bid, 5, 3),
    new Order(OrderType.bid, 4, 3),
    new Order(OrderType.bid, 3, 3),
    new Order(OrderType.bid, 2, 3),
    new Order(OrderType.bid, 1, 3)
];

describe('When rendering the OrderBook component', () => {
    describe('without a websocket connection', () => {
        it('shows the OrderBook title, Depth, Render throttle and Toggle Feed elements', async () => {
            render(<OrderBook defaultDepth={15} throttle={350} />);

            // query* functions will return the element or null if it cannot be found
            // get* functions will return the element or throw an error if it cannot be found
            expect(screen.queryByText('XBTUSD')).toBeInTheDocument();
            expect(screen.queryByText('Depth')).toBeInTheDocument();
            expect(screen.queryByText('Render throttle')).toBeInTheDocument();
            expect(screen.queryByText('Toggle Feed')).toBeInTheDocument();

            expect(screen.queryByRole("button", { id: /toggleFeed/i })).toBeInTheDocument();
            expect(screen.queryAllByRole('slider').length).toBe(2);

            expect(screen.queryByText('Spread')).not.toBeInTheDocument();
        });

        it('Shows the No websocket connection made', async () => {
            render(<OrderBook defaultDepth={15} throttle={350} />);

            expect(screen.queryByText('No websocket connection')).not.toBeNull();
        });
    });


    describe('with a websocket connection', () => {
        it('shows the OrderBook title, Depth, Render throttle and Toggle Feed elements', async () => {
            jest.spyOn(hooks, "useOrderBookFeed").mockImplementation(() => {
                return {
                    wasDisconnected: false,
                    reconnectHandler: jest.fn(),
                    asks: asks,
                    bids: bids,
                    pair: "PI_XBTUSD",
                    handleFeedToggle: jest.fn()
                };
            });

            render(<OrderBook defaultDepth={15} throttle={350} />);

            expect(screen.queryByText('XBTUSD')).toBeInTheDocument();
            expect(screen.queryByText('Depth')).toBeInTheDocument();
            expect(screen.queryByText('Render throttle')).toBeInTheDocument();
            expect(screen.queryByText('Toggle Feed')).toBeInTheDocument();

            expect(screen.getByRole("button", { id: /toggleFeed/i })).toBeInTheDocument();
            expect(screen.getAllByRole('slider').length).toBe(2);
        });


        it('does not show the No websocket connection made', async () => {
            jest.spyOn(hooks, "useOrderBookFeed").mockImplementation(() => {
                return {
                    wasDisconnected: false,
                    reconnectHandler: jest.fn(),
                    asks: asks,
                    bids: bids,
                    pair: "PI_XBTUSD",
                    handleFeedToggle: jest.fn()
                };
            });

            render(<OrderBook defaultDepth={15} throttle={350} />)

            expect(screen.queryByText('No websocket connection')).not.toBeInTheDocument();
        });

        it('shows all expected rows and headers', async () => {
            jest.spyOn(hooks, "useOrderBookFeed").mockImplementation(() => {
                return {
                    wasDisconnected: false,
                    reconnectHandler: jest.fn(),
                    asks: asks,
                    bids: bids,
                    pair: "PI_XBTUSD",
                    handleFeedToggle: jest.fn()
                };
            });

            render(<OrderBook defaultDepth={15} throttle={350} />)

            expect(screen.queryAllByText('PRICE').length).toBe(2);
            expect(screen.queryAllByText('SIZE').length).toBe(2);
            expect(screen.queryAllByText('TOTAL').length).toBe(2);

            expect(screen.queryByText('Spread: 1.0 (6.67%)')).toBeInTheDocument();

            asks.forEach(ask => expect(screen.queryAllByText(ask.price.toFixed(2).toLocaleString()).length).toBeGreaterThan(0));
            bids.forEach(bid => expect(screen.queryAllByText(bid.price.toFixed(2).toLocaleString()).length).toBeGreaterThan(0));
        });

        it('shows the reconnect modal when the websocket says it disconnects', async () => {
            jest.spyOn(hooks, "useOrderBookFeed").mockImplementation(() => {
                return {
                    wasDisconnected: true,
                    reconnectHandler: jest.fn(),
                    asks: asks,
                    bids: bids,
                    pair: "PI_XBTUSD",
                    handleFeedToggle: jest.fn()
                };
            });

            render(<OrderBook defaultDepth={15} throttle={350} />)

            expect(screen.queryByText('Websocket disconnected, click to reconnect')).toBeInTheDocument();
            expect(screen.queryByText("Reconnect")).toBeInTheDocument();
        });
    });
});