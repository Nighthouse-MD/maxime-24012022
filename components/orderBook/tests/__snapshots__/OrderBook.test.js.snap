// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When rendering the OrderBook component without a websocket connection renders correctly for OrderType.ask 1`] = `
<div
  className="orderBookContainer container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h2>
        XBTUSD
      </h2>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="row"
      >
        <div
          className="settingsLabelCol col-4"
        >
          Depth
        </div>
        <div
          className="settingsInputCol col-4"
        >
          <input
            className="settingsInput"
            id="depth"
            max="40"
            min="0"
            onChange={[Function]}
            type="range"
            value={15}
          />
        </div>
        <div
          className="settingsValueCol col-4"
        >
          15
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="settingsLabelCol col-4"
        >
          Render throttle
        </div>
        <div
          className="settingsInputCol col-4"
        >
          <input
            className="settingsInput"
            id="renderThrottle"
            max="2000"
            min="0"
            onChange={[Function]}
            type="range"
            value={350}
          />
        </div>
        <div
          className="settingsValueCol col-4"
        >
          350
           ms
        </div>
      </div>
    </div>
  </div>
  <hr />
  <br />
  <div
    className="row"
  >
    <div
      className="orderBookTitle col"
    >
      Order book
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div
        className="row"
      >
        <div
          className="noConnection col"
        >
          No websocket connection
        </div>
      </div>
    </div>
  </div>
  <br />
  <div
    className="row"
  >
    <div
      className="col"
    >
      <button
        className="toggleButton btn btn-primary btn-sm"
        disabled={false}
        id="toggleFeed"
        name="toggleFeed"
        onClick={[Function]}
        type="button"
      >
        Toggle Feed
      </button>
    </div>
  </div>
</div>
`;
